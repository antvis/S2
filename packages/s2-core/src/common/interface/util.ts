export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
