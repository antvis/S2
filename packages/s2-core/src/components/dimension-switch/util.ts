import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
