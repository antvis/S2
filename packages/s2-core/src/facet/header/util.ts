import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
