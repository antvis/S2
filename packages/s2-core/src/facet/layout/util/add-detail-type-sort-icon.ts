import { Group } from '@antv/g-canvas';
import * as _ from 'lodash';
import { GuiIcon } from '../../../common/icons';
import { DEFAULT_PADDING, KEY_LIST_SORT } from '../../../common/constant';
import { SpreadSheet } from '../../..';

// 排序按钮的宽度
export const SORT_ICON_SIZE = 8;

export const getIconType = (key: string, spreadsheet: SpreadSheet) => {
  /*
  currentSortKey 存储的点击的某个field对应的升序还是降序
  currentSortKey = {
   [field]: up/down
  }
   */
  const currentSortKey = spreadsheet.store.get('currentSortKey', {});
  let upIconType = 'SortUp';
  let downIconType = 'SortDown';
  if (currentSortKey && _.has(currentSortKey, key)) {
    // 有配置,当前点击的过的key(某个维度)
    if (_.get(currentSortKey, key) === 'up') {
      // 点击过此维度的up
      upIconType = 'SortUpSelected';
    } else {
      // 点击过此维度的down
      downIconType = 'SortDownSelected';
    }
  }

  return {
    upIconType,
    downIconType,
  };
};

export const renderIcon = (
  parent: Group,
  spreadsheet: SpreadSheet,
  x: number,
  y: number,
  type: string,
  key: string,
  sortKeyVal: string,
  sortType: string,
) => {
  const icon = new GuiIcon({
    type,
    x,
    y,
    width: SORT_ICON_SIZE,
    height: SORT_ICON_SIZE,
  });

  icon.on('click', () => {
    const selected = icon.get('selected');
    let currentSortKey = {};
    if (!selected) {
      currentSortKey = {
        [key]: sortKeyVal,
      };
    }
    spreadsheet.store.set('currentSortKey', currentSortKey);
    icon.set('selected', !selected);
    spreadsheet.emit(KEY_LIST_SORT, {
      sortFieldId: key,
      sortMethod: selected ? '' : sortType,
    });
    spreadsheet.render(false);
  });

  parent.add(icon);
};

export const renderDetailTypeSortIcon = (
  parent: Group,
  spreadsheet: SpreadSheet,
  textX: number,
  textY,
  key,
) => {
  const iconType = getIconType(key, spreadsheet);

  renderIcon(
    parent,
    spreadsheet,
    textX,
    textY - DEFAULT_PADDING * 2,
    iconType.upIconType,
    key,
    'up',
    'ASC',
  );
  renderIcon(
    parent,
    spreadsheet,
    textX,
    textY - DEFAULT_PADDING / 2,
    iconType.downIconType,
    key,
    'down',
    'DESC',
  );
};
