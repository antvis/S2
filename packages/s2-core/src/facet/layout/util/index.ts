export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
