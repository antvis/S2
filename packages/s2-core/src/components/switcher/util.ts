import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
