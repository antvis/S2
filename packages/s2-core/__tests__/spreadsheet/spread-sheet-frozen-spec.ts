import { map, merge } from 'lodash';
import { customRowGridFields } from 'tests/data/custom-grid-fields';
import { CustomGridData } from 'tests/data/data-custom-grid';
import * as mockDataConfig from 'tests/data/mock-dataset.json';
import { getContainer } from 'tests/util/helpers';
import {
  PivotSheet,
  type FrozenFacet,
  type S2DataConfig,
  type S2Options,
  type SpreadSheet,
} from '../../src';
import { pickMap } from '../util/fp';

function expectFrozenGroup(
  s2: PivotSheet,
  headerName: 'rowHeader' | 'columnHeader' | 'seriesNumberHeader' = 'rowHeader',
) {
  const pickCoordinate = pickMap(['id', 'x', 'y', 'width', 'height']);

  const actualHead = pickCoordinate(
    map(s2.facet[headerName]?.frozenGroup.children, 'meta'),
  );

  expect(actualHead).toMatchSnapshot();

  const actualTrailing = pickCoordinate(
    map(s2.facet[headerName]?.frozenTrailingGroup.children, 'meta'),
  );

  expect(actualTrailing).toMatchSnapshot();
}

function getFrozenGroupPosition(
  s2: PivotSheet,
  headerName: 'rowHeader' | 'columnHeader' = 'rowHeader',
) {
  return [
    s2.facet[headerName]?.frozenGroup.getPosition().map(Math.floor),
    s2.facet[headerName]?.frozenTrailingGroup.getPosition().map(Math.floor),
  ];
}

describe('Spread Sheet Frozen Tests', () => {
  let container: HTMLElement;

  const scrollX = (s2: SpreadSheet, value: number) => {
    s2.interaction.scrollTo({
      offsetX: { value, animate: false },
      offsetY: { value, animate: false },
      rowHeaderOffsetX: { value, animate: false },
    });
  };

  beforeEach(() => {
    container = getContainer();
  });

  afterEach(() => {
    container?.remove();
  });

  describe('grid mode', () => {
    test('should render correct frozen areas when has vertical scroll area', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 300,
        hierarchyType: 'grid',
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 0,
          range: [0, 1],
        },
        frozenTrailingRow: {
          height: 30,
          y: 210,
          range: [7, 7],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct frozen areas when has no vertical scroll area', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 800,
        hierarchyType: 'grid',
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 0,
          range: [0, 1],
        },
        frozenTrailingRow: {
          height: 30,
          y: 210,
          range: [7, 7],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct frozen areas with pagination', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 300,
        hierarchyType: 'grid',
        pagination: {
          current: 2,
          pageSize: 4,
        },
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 120,
          range: [4, 5],
        },
        frozenTrailingRow: {
          height: 30,
          y: 210,
          range: [7, 7],
        },
      });
      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct series frozen areas', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 300,
        hierarchyType: 'grid',
        seriesNumber: {
          enable: true,
        },
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expectFrozenGroup(s2, 'seriesNumberHeader');

      s2.destroy();
    });
  });

  describe('tree mode', () => {
    test('should render correct frozen areas when has vertical scroll area', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 300,
        hierarchyType: 'tree',
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 0,
          range: [0, 1],
        },
        frozenTrailingRow: {
          height: 30,
          y: 270,
          range: [9, 9],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct frozen areas when has no vertical scroll area', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 800,
        hierarchyType: 'tree',
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 0,
          range: [0, 1],
        },
        frozenTrailingRow: {
          height: 30,
          y: 270,
          range: [9, 9],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct frozen areas with pagination', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 300,
        hierarchyType: 'tree',
        pagination: {
          current: 2,
          pageSize: 4,
        },
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 120,
          range: [4, 5],
        },
        frozenTrailingRow: {
          height: 30,
          y: 210,
          range: [7, 7],
        },
      });
      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct series frozen areas', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, {
        width: 300,
        height: 300,
        hierarchyType: 'tree',
        seriesNumber: {
          enable: true,
        },
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expectFrozenGroup(s2, 'seriesNumberHeader');

      s2.destroy();
    });
  });

  describe('custom tree mode', () => {
    const pivotSheetCustomRowGridDataCfg: S2DataConfig = {
      data: CustomGridData,
      meta: [
        ...mockDataConfig.meta,
        {
          field: 'a-1',
          name: '层级1',
        },
        {
          field: 'a-1-1',
          name: '层级2',
        },
        {
          field: 'measure-1',
          name: '层级3',
        },
        {
          field: 'measure-1',
          formatter: (value) => `#-${value}`,
        },
      ],
      fields: customRowGridFields,
    };

    test('should render correct frozen areas for custom grid mode', async () => {
      const s2 = new PivotSheet(container, pivotSheetCustomRowGridDataCfg, {
        width: 600,
        height: 400,
        hierarchyType: 'grid',
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 0,
          range: [0, 1],
        },
        frozenTrailingRow: {
          height: 30,
          y: 240,
          range: [8, 8],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct frozen areas for custom tree mode', async () => {
      const s2 = new PivotSheet(container, pivotSheetCustomRowGridDataCfg, {
        width: 600,
        height: 400,
        hierarchyType: 'tree',
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 0,
          range: [0, 1],
        },
        frozenTrailingRow: {
          height: 30,
          y: 420,
          range: [14, 14],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct frozen areas for custom grid mode with pagination', async () => {
      const s2 = new PivotSheet(container, pivotSheetCustomRowGridDataCfg, {
        width: 600,
        height: 400,
        hierarchyType: 'grid',
        pagination: {
          current: 2,
          pageSize: 4,
        },
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenRow: {
          height: 60,
          y: 120,
          range: [4, 5],
        },
        frozenTrailingRow: {
          height: 30,
          y: 210,
          range: [7, 7],
        },
      });

      expectFrozenGroup(s2);
      s2.destroy();
    });

    test('should render correct series frozen areas', async () => {
      const s2 = new PivotSheet(container, pivotSheetCustomRowGridDataCfg, {
        width: 300,
        height: 300,
        hierarchyType: 'tree',
        seriesNumber: {
          enable: true,
        },
        frozen: {
          rowCount: 2,
          trailingRowCount: 1,
        },
      });

      await s2.render();

      expectFrozenGroup(s2, 'seriesNumberHeader');

      s2.destroy();
    });
  });

  describe('pivot col header frozen', () => {
    const baseOptions: S2Options = {
      width: 600,
      height: 400,
      frozen: {
        colCount: 1,
        trailingColCount: 1,
      },
      style: {
        colCell: {
          width: 100,
          widthByField: {
            'root[&]家具[&]沙发[&]number': 400,
            'root[&]办公用品[&]笔[&]number': 400,
          },
        },
      },
    };

    test('should render correct frozen areas for row header frozen', async () => {
      const s2 = new PivotSheet(container, mockDataConfig, baseOptions);

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenCol: {
          width: 100,
          x: 0,
          range: [0, 0],
        },
        frozenTrailingCol: {
          width: 100,
          x: 900,
          range: [3, 3],
        },
      });

      expectFrozenGroup(s2, 'columnHeader');

      const prev = getFrozenGroupPosition(s2, 'columnHeader');

      scrollX(s2, 100);

      // 移动后，frozen col 和 trailing col 的位置都不变
      expect(getFrozenGroupPosition(s2, 'columnHeader')).toEqual(prev);

      s2.destroy();
    });

    test('should render correct frozen areas for row header is not frozen', async () => {
      const s2 = new PivotSheet(
        container,
        mockDataConfig,
        merge({}, baseOptions, {
          frozen: {
            rowHeader: false,
          },
        }),
      );

      await s2.render();

      expect((s2.facet as FrozenFacet).frozenGroupAreas).toMatchObject({
        frozenCol: {
          width: 100,
          x: 0,
          range: [0, 0],
        },
        frozenTrailingCol: {
          width: 100,
          x: 900,
          range: [3, 3],
        },
      });

      expectFrozenGroup(s2, 'columnHeader');
      let prev = getFrozenGroupPosition(s2, 'columnHeader');

      scrollX(s2, 100);

      // 移动后，frozen col 会改变 而 trailing col 的位置不变
      let current = getFrozenGroupPosition(s2, 'columnHeader');

      expect(current[0]?.[0]).toEqual(prev[0]?.[0]! - 100);
      expect(current[1]).toEqual(prev[1]);

      // 移动超过角头宽度
      // 移动后，frozen col 和 trailing col 的位置都不变
      scrollX(s2, 300);

      prev = getFrozenGroupPosition(s2, 'columnHeader');

      scrollX(s2, 300);

      current = getFrozenGroupPosition(s2, 'columnHeader');

      expect(current).toEqual(prev);
      expect(current).toEqual([
        [2, 0, 0],
        [-400, 0, 0],
      ]);

      s2.destroy();
    });
  });
});
