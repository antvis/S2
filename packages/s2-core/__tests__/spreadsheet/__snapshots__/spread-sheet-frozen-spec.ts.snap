// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spread Sheet Frozen Tests custom tree mode should render correct frozen areas for custom grid mode 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]a-1[&]a-1-1[&]measure-1",
    "width": 99.66,
    "x": 199.32,
    "y": 0,
  },
  Object {
    "height": 30,
    "id": "root[&]a-1[&]a-1-1[&]a-1-1-1",
    "width": 99.66,
    "x": 199.32,
    "y": 30,
  },
  Object {
    "height": 60,
    "id": "root[&]a-1[&]a-1-1",
    "width": 99.66,
    "x": 99.66,
    "y": 0,
  },
  Object {
    "height": 60,
    "id": "root[&]a-1",
    "width": 99.66,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct frozen areas for custom grid mode 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]a-3[&]measure-5",
    "width": 199.32,
    "x": 99.66,
    "y": 240,
  },
  Object {
    "height": 30,
    "id": "root[&]a-3",
    "width": 99.66,
    "x": 0,
    "y": 240,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct frozen areas for custom grid mode with pagination 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]a-2[&]a-2-1",
    "width": 199.32,
    "x": 99.66,
    "y": 120,
  },
  Object {
    "height": 30,
    "id": "root[&]a-2[&]a-2-2[&]measure-3",
    "width": 99.66,
    "x": 199.32,
    "y": 150,
  },
  Object {
    "height": 60,
    "id": "root[&]a-2",
    "width": 99.66,
    "x": 0,
    "y": 120,
  },
  Object {
    "height": 30,
    "id": "root[&]a-2[&]a-2-2",
    "width": 99.66,
    "x": 99.66,
    "y": 150,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct frozen areas for custom grid mode with pagination 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]a-3[&]a-3-2[&]measure-4",
    "width": 99.66,
    "x": 199.32,
    "y": 210,
  },
  Object {
    "height": 30,
    "id": "root[&]a-3[&]a-3-2",
    "width": 99.66,
    "x": 99.66,
    "y": 210,
  },
  Object {
    "height": 30,
    "id": "root[&]a-3",
    "width": 99.66,
    "x": 0,
    "y": 210,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct frozen areas for custom tree mode 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]a-1",
    "width": 271.03125,
    "x": 0,
    "y": 0,
  },
  Object {
    "height": 30,
    "id": "root[&]a-1[&]a-1-1",
    "width": 271.03125,
    "x": 0,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]a-1",
    "width": 271.03125,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct frozen areas for custom tree mode 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]a-3[&]measure-5",
    "width": 271.03125,
    "x": 0,
    "y": 420,
  },
  Object {
    "height": -90,
    "id": "root[&]a-3",
    "width": 271.03125,
    "x": 0,
    "y": 420,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct series frozen areas 1`] = `
Array [
  Object {
    "height": 60,
    "id": "1",
    "width": 80,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests custom tree mode should render correct series frozen areas 2`] = `
Array [
  Object {
    "height": 30,
    "id": "3",
    "width": 80,
    "x": 0,
    "y": 420,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct frozen areas when has no vertical scroll area 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]杭州市",
    "width": 96,
    "x": 96,
    "y": 0,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]绍兴市",
    "width": 96,
    "x": 96,
    "y": 30,
  },
  Object {
    "height": 60,
    "id": "root[&]浙江省",
    "width": 96,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct frozen areas when has no vertical scroll area 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]乐山市",
    "width": 96,
    "x": 96,
    "y": 210,
  },
  Object {
    "height": 30,
    "id": "root[&]四川省",
    "width": 96,
    "x": 0,
    "y": 210,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct frozen areas when has vertical scroll area 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]杭州市",
    "width": 96,
    "x": 96,
    "y": 0,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]绍兴市",
    "width": 96,
    "x": 96,
    "y": 30,
  },
  Object {
    "height": 60,
    "id": "root[&]浙江省",
    "width": 96,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct frozen areas when has vertical scroll area 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]乐山市",
    "width": 96,
    "x": 96,
    "y": 210,
  },
  Object {
    "height": 30,
    "id": "root[&]四川省",
    "width": 96,
    "x": 0,
    "y": 210,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct frozen areas with pagination 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]成都市",
    "width": 96,
    "x": 96,
    "y": 120,
  },
  Object {
    "height": 30,
    "id": "root[&]四川省[&]绵阳市",
    "width": 96,
    "x": 96,
    "y": 150,
  },
  Object {
    "height": 60,
    "id": "root[&]四川省",
    "width": 96,
    "x": 0,
    "y": 120,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct frozen areas with pagination 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]乐山市",
    "width": 96,
    "x": 96,
    "y": 210,
  },
  Object {
    "height": 30,
    "id": "root[&]四川省",
    "width": 96,
    "x": 0,
    "y": 210,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct series frozen areas 1`] = `
Array [
  Object {
    "height": 60,
    "id": "1",
    "width": 80,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests grid mode should render correct series frozen areas 2`] = `
Array [
  Object {
    "height": 30,
    "id": "2",
    "width": 80,
    "x": 0,
    "y": 210,
  },
]
`;

exports[`Spread Sheet Frozen Tests pivot col header frozen should render correct frozen areas for row header frozen 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]家具[&]桌子[&]number",
    "width": 100,
    "x": 0,
    "y": 60,
  },
  Object {
    "height": 30,
    "id": "root[&]家具[&]桌子",
    "width": 100,
    "x": 0,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]家具",
    "width": 100,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests pivot col header frozen should render correct frozen areas for row header frozen 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]办公用品[&]纸张[&]number",
    "width": 100,
    "x": 900,
    "y": 60,
  },
  Object {
    "height": 30,
    "id": "root[&]办公用品[&]纸张",
    "width": 100,
    "x": 900,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]办公用品",
    "width": 100,
    "x": 900,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests pivot col header frozen should render correct frozen areas for row header is not frozen 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]家具[&]桌子[&]number",
    "width": 100,
    "x": 0,
    "y": 60,
  },
  Object {
    "height": 30,
    "id": "root[&]家具[&]桌子",
    "width": 100,
    "x": 0,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]家具",
    "width": 100,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests pivot col header frozen should render correct frozen areas for row header is not frozen 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]办公用品[&]纸张[&]number",
    "width": 100,
    "x": 900,
    "y": 60,
  },
  Object {
    "height": 30,
    "id": "root[&]办公用品[&]纸张",
    "width": 100,
    "x": 900,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]办公用品",
    "width": 100,
    "x": 900,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct frozen areas when has no vertical scroll area 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]浙江省",
    "width": 120,
    "x": 0,
    "y": 0,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]杭州市",
    "width": 120,
    "x": 0,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江省",
    "width": 120,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct frozen areas when has no vertical scroll area 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]乐山市",
    "width": 120,
    "x": 0,
    "y": 270,
  },
  Object {
    "height": -90,
    "id": "root[&]四川省",
    "width": 120,
    "x": 0,
    "y": 270,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct frozen areas when has vertical scroll area 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]浙江省",
    "width": 120,
    "x": 0,
    "y": 0,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]杭州市",
    "width": 120,
    "x": 0,
    "y": 30,
  },
  Object {
    "height": 30,
    "id": "root[&]浙江省",
    "width": 120,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct frozen areas when has vertical scroll area 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]乐山市",
    "width": 120,
    "x": 0,
    "y": 270,
  },
  Object {
    "height": -90,
    "id": "root[&]四川省",
    "width": 120,
    "x": 0,
    "y": 270,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct frozen areas with pagination 1`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]浙江省[&]舟山市",
    "width": 120,
    "x": 0,
    "y": 120,
  },
  Object {
    "height": 30,
    "id": "root[&]四川省",
    "width": 120,
    "x": 0,
    "y": 150,
  },
  Object {
    "height": -90,
    "id": "root[&]浙江省",
    "width": 120,
    "x": 0,
    "y": 120,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct frozen areas with pagination 2`] = `
Array [
  Object {
    "height": 30,
    "id": "root[&]四川省[&]绵阳市",
    "width": 120,
    "x": 0,
    "y": 210,
  },
  Object {
    "height": -30,
    "id": "root[&]四川省",
    "width": 120,
    "x": 0,
    "y": 210,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct series frozen areas 1`] = `
Array [
  Object {
    "height": 60,
    "id": "1",
    "width": 80,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`Spread Sheet Frozen Tests tree mode should render correct series frozen areas 2`] = `
Array [
  Object {
    "height": 30,
    "id": "2",
    "width": 80,
    "x": 0,
    "y": 270,
  },
]
`;
